geps/gep-1742.md

```diff
@@ -301,8 +301,134 @@ Could not find any HTTP specific timeouts. PRs welcomed. 😊
 
 ## API
 
-TBD.
+The above diagrams show that there are many different kinds of configurable timeouts
+supported by Gateway implementations: connect, idle, request, upstream, downstream.
+Although there may be opportunity for the specification of a common API for more of
+them in the future, this GEP will focus on the L7 timeouts in HTTPRoutes that are
+most valuable to clients.
+
+From the above analysis, it appears that, at a minimum, implementations are capable of
+supporting the configuration of simple client downstream request timeouts on HTTPRoute
+rules. This is a relatively small addition that would benefit many users.
+
+A client HTTP request to a gateway may result in more than one call to the destination
+backend service, for example, if automatic retries are supported. Some implementations
+support configuring a timeout for the individual backend requests, seperate from the
+overall client request timeout. Adding non-Core support for this would also benefit
+many users.
+
+### GO
+
+```go
+type HTTPRouteRule struct {
+	// Timeouts defines the timeouts that can be configured for an HTTP request.
+	//
+	// Support: Core
+	//
+	// +optional
+	// <gateway:experimental>
+	Timeouts *HTTPRouteTimeouts `json:"timeouts,omitempty"`
+
+	// ...
+}
+
+// HTTPRouteTimeouts defines timeouts that can be configured for an HTTPRoute.
+// Timeout values are formatted like 1h/1m/1s/1ms as parsed by Golang time.ParseDuration
+// and MUST BE >= 1ms.
+type HTTPRouteTimeouts struct {
+	// Request specifies a timeout for responding to client HTTP requests,
+	// disabled by default.
+	//
+	// For example, the following rule will timeout if a client request is taking
+	// longer than 10 seconds to complete:
+	//
+	// ```
+	// rules:
+	// - timeouts:
+	//     request: 10s
+	//   backendRefs:
+	//     ...
+	// ```
+	//
+	// Support: Core
+	//
+	// +optional
+	Request *metav1.Duration `json:"request,omitempty"`
+
+	// BackendRequest specifies a timeout for an individual request from the gateway
+	// to a backend service.
+	// Default is the value of Request timeout.
+	//
+	// Support: Extended
+	//
+	// +optional
+	BackendRequest *metav1.Duration `json:"backendRequest,omitempty"`
+}
+```
+
+### YAML
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: timeout-example
+spec:
+  ...
+  rules:
+  - backendRefs:
+    - name: some-service
+      port: 8080
+    timeouts:
+      request: 10s
+      backendRequest: 2s
+```
+
+### Timeout values
+
+There a 2 kinds of timeouts that can be configured in an `HTTPRouteRule`:
+
+1. `timeouts.request` is the timeout for the Gateway API implementation to send a
+    response to a client HTTP request. Whether the gateway starts the timeout before
+    or after the entire client request stream has been received, is implementation dependent.
+    This field is required `Core` support.
```

Comment:
Usually fields marked as `Core` support have tighter requirements than this and don't have implementation-specific components. I think this might be OK, but we need to be confident that we can write a reliable conformance test for this that would be compatible with all implementations. Are we sure that _all_ implementations can support this?

Reply:
My first thought was to make it a requirement that the timeout start after the client request has been fully received, @youngnick had the same thought apparently (https://github.com/kubernetes-sigs/gateway-api/pull/1997#discussion_r1195445362, but unfortunately it seems that some proxies can't do that so this is the compromise.

I think a conformance test compatible with all instances is one that simply checks that a request times out when a backend is very slow, which btw is the main motivating use case for this feature.

Reply:
> I think a conformance test compatible with all instances is one that simply checks that a request times out when a backend is very slow, which btw is the main motivating use case for this feature.

rereading this intent, I'm not sure the description and naming of this field really line up? may be nitpicky and discussed before but did we consider naming this something like `backendRequest` if the intent is to cover the case that a backend is slow seems like the naming should reflect that

`the timeout for the Gateway API implementation to send a response to a client HTTP request` sounds like something the client itself should be enforcing (as it is written here)

Reply:
> did we consider naming this something like `backendRequest`

This proposal actually has an optional non-Core field named `backendRequest` 😃 It's used to set a timeout for a single request from the gateway to the backend. The `request` timeout is controlling how long the client should wait for a response given that the gateway may be making multiple backend requests (e.g., retries) before replying. Note that without retries or other fancy things going on, the two timeouts degenerate to more-or-less the same thing.

Reply:
ah right, confused that field since it was talking about retries as well so forgot about it, but that field sounds to me to cover more like what the intent of your comment about testing above is

the language `The request timeout is controlling how long the client should wait for a response...` similarly `timeouts.request is the timeout for the Gateway API implementation to send a response to a client HTTP request` sounds like a timeout that a client would enforce

If the field is documented like this but really is implemented differently/means something else I think this will be confusing for users, implementers, and those trying to give advice on supporting an implementation (we already have trouble with users of Contour getting confused about timeouts we expose, which are a lot more specific in that they tie in to Envoy timeouts and cover specific hops in the request flow)

(this and other threads may be rehashing why this wasn't done earlier, generalizing timeouts feels like it will lead to confusion)

Reply:
another general option as opposed to trying to generalize timeouts would be to make timeouts for all hops of a proxied request "extended" support and implementations could advertise what they support via conformance profile reporting as well as documentation

Reply:
I was trying to describe `timeout.request` in a way to be clear that it is `for the Gateway API implementation` to timeout doing it's job (which mostly means while waiting for 1 or more backend requests to finish). Maybe there's a better way to word it? I'll give it some thought, and suggestions are welcome 😃

Reply:
> which has advantages but also a lot of disadvantages tbh

Yeah, it would be nice to make at least something Core.

Reply:
Updated the comment in the Go struct for more clarity.